File: frontend/smart-devices-frontend/src/app/models/device.ts
import { Comment } from './comment';

export interface Device {
  id?: number;
  name: string;
  brand: string;
  type: DeviceType;
  description?: string;
  price: number;
  releaseDate?: string;
  imageUrl?: string;
  specifications?: string;
  comments?: Comment[];
}

export enum DeviceType {
  SMARTPHONE = 'SMARTPHONE',
  LAPTOP = 'LAPTOP',
  TABLET = 'TABLET',
  SMARTWATCH = 'SMARTWATCH',
  HEADPHONES = 'HEADPHONES',
  SMART_SPEAKER = 'SMART_SPEAKER',
  CAMERA = 'CAMERA',
  GAMING_CONSOLE = 'GAMING_CONSOLE'
}

export const DeviceTypeLabels: Record<DeviceType, string> = {
  [DeviceType.SMARTPHONE]: 'Smartphone',
  [DeviceType.LAPTOP]: 'Laptop',
  [DeviceType.TABLET]: 'Tablet',
  [DeviceType.SMARTWATCH]: 'Smartwatch',
  [DeviceType.HEADPHONES]: 'Headphones',
  [DeviceType.SMART_SPEAKER]: 'Smart Speaker',
  [DeviceType.CAMERA]: 'Camera',
  [DeviceType.GAMING_CONSOLE]: 'Gaming Console'
};

File: src/main/java/com/smartdevices/model/Device.java
package com.smartdevices.model;

import com.fasterxml.jackson.annotation.JsonManagedReference;
import jakarta.persistence.*;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotNull;
import jakarta.validation.constraints.Positive;
import java.math.BigDecimal;
import java.time.LocalDate;
import java.util.ArrayList;
import java.util.List;

/**
 * Entity representing a smart device
 * Part of the Model layer in layered architecture
 */
@Entity
@Table(name = "devices")
public class Device {
    
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;
    
    @NotBlank(message = "Name is required")
    @Column(nullable = false)
    private String name;
    
    @NotBlank(message = "Brand is required")
    @Column(nullable = false)
    private String brand;
    
    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private DeviceType type;
    
    @Column(length = 1000)
    private String description;
    
    @NotNull(message = "Price is required")
    @Positive(message = "Price must be positive")
    @Column(nullable = false)
    private BigDecimal price;
    
    @Column(name = "release_date")
    private LocalDate releaseDate;
    
    @Column(name = "image_url")
    private String imageUrl;
    
    @Column(name = "specifications", length = 2000)
    private String specifications;
    
    @OneToMany(mappedBy = "device", cascade = CascadeType.ALL, fetch = FetchType.LAZY)
    @JsonManagedReference
    private List<Comment> comments = new ArrayList<>();
    
    // Constructors
    public Device() {}
    
    public Device(String name, String brand, DeviceType type, BigDecimal price) {
        this.name = name;
        this.brand = brand;
        this.type = type;
        this.price = price;
    }
    
    // Getters and Setters
    public Long getId() { return id; }
    public void setId(Long id) { this.id = id; }
    
    public String getName() { return name; }
    public void setName(String name) { this.name = name; }
    
    public String getBrand() { return brand; }
    public void setBrand(String brand) { this.brand = brand; }
    
    public DeviceType getType() { return type; }
    public void setType(DeviceType type) { this.type = type; }
    
    public String getDescription() { return description; }
    public void setDescription(String description) { this.description = description; }
    
    public BigDecimal getPrice() { return price; }
    public void setPrice(BigDecimal price) { this.price = price; }
    
    public LocalDate getReleaseDate() { return releaseDate; }
    public void setReleaseDate(LocalDate releaseDate) { this.releaseDate = releaseDate; }
    
    public String getImageUrl() { return imageUrl; }
    public void setImageUrl(String imageUrl) { this.imageUrl = imageUrl; }
    
    public String getSpecifications() { return specifications; }
    public void setSpecifications(String specifications) { this.specifications = specifications; }
    
    public List<Comment> getComments() { return comments; }
    public void setComments(List<Comment> comments) { this.comments = comments; }
    
    // Helper methods
    public void addComment(Comment comment) {
        comments.add(comment);
        comment.setDevice(this);
    }
    
    public void removeComment(Comment comment) {
        comments.remove(comment);
        comment.setDevice(null);
    }
}
